### PR TITLE
Exchange

### PR DESCRIPTION
**Pull request by leonidstr**

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Helper2FA.java [hunk 1]
```diff
@@ -154,32 +141,46 @@ public static Status2FA auth2FA(String secretPhrase, int code) throws ParameterE
...
     }
 
-    private static void validateKeyStoreStatus(long accountId, VaultKeyStore.Status status, String notPerformedAction) throws ParameterException {
-        if (status != VaultKeyStore.Status.OK) {
+    private static void validateKeyStoreStatus(long accountId, KeyStoreService.Status status, String notPerformedAction) throws ParameterException {
+        if (status != KeyStoreService.Status.OK) {
             LOG.debug( "Vault wallet not " + notPerformedAction + " {} - {}", Convert2.rsAccount(accountId), status);
             throw new ParameterException("Unable to generate account", null, JSONResponses.vaultWalletError(accountId, notPerformedAction,
                     status.message));
         }
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/Helper2FA.java]: what should happen in ERROR case ?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/VaultKeyStoreServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+        this(keystoreDir, CURRENT_KEYSTORE_VERSION);
+    }
+
+     public VaultKeyStoreServiceImpl(Path keystoreDir, Integer version) {
+         if (version < 0) {
+             throw new IllegalArgumentException("version should not be negative");
+         }
+         this.version = version;
+         this.keystoreDirPath = keystoreDir;
+         this.ntpTime = CDI.current().select(NtpTime.class).get();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/VaultKeyStoreServiceImpl.java]: Move to @Inject constructor

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/ParameterParser.java [hunk 3]
```diff
@@ -604,6 +607,12 @@ public static String getPassphrase(HttpServletRequest req, boolean isMandatory)
         String secretPhrase = getStringParameter(req, "passphrase", isMandatory);
         return API.elGamalDecrypt(secretPhrase);
     }
+    public static String getPassphrase(String passphrase, boolean isMandatory) throws ParameterException {
+        if (StringUtils.isEmpty(passphrase) && isMandatory) {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/ParameterParser.java]: Use apl-util StringUtils

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/eth/service/EthereumWalletService.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+public class EthereumWalletService {
+    private static final Logger log = getLogger(EthereumWalletService.class);
+
+    private Web3j web3j = CDI.current().select(Web3j.class).get();
+    private PropertiesHolder propertiesHolder = CDI.current().select(PropertiesHolder.class).get();
+    private final KeyStoreService keyStoreService = CDI.current().select(KeyStoreService.class).get();
+
+    private String paxContractAddress = propertiesHolder.getStringProperty("apl.eth.pax.contract.address");
+
+    public BigInteger getPaxBalanceWei(String accountAddress){
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/eth/service/EthereumWalletService.java]: No NULL checks on methods, NPE is possible...

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import javax.inject.Singleton;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+
+@Singleton
+public class DexService {
+    private static final Logger LOG = LoggerFactory.getLogger(DexService.class);
+
+    private EthereumWalletService ethereumWalletService = CDI.current().select(EthereumWalletService.class).get();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java]: Move to @Injectable constructor

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/KeyStoreServiceTest.java [hunk 6]
```diff
@@ -38,7 +38,7 @@
 import org.mockito.Mockito;
 
 @EnableWeld
-public class VaultKeyStoreTest {
+public class KeyStoreServiceTest {
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/KeyStoreServiceTest.java]: Add TODO to rewrite this test

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/eth/service/EthereumWalletService.java [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+    public TransactionReceipt waitForReceipt(String transactionHash) {
+        int attempts = Web3jUtils.CONFIRMATION_ATTEMPTS;
+        int sleep_millis = Web3jUtils.SLEEP_DURATION;
+
+        Optional<TransactionReceipt> receipt = getReceipt(transactionHash);
+
+        while(attempts-- > 0 && !receipt.isPresent()) {
+            try {
+                Thread.sleep(sleep_millis);
```
[the remark this answers is not in the document]
Comment (leonidstr) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/eth/service/EthereumWalletService.java]: It's just a wrapper above Thread.sleep

**Pull request merged**